### PR TITLE
test: a pytest script to generate significant load on a FT contract

### LINKED ISSUE #8487: Implement a benchmark to estimate fungible token transfer throughput under the real-world conditions
We want to know how long on average a continuous load of blocks with 100 fungible token transfers each takes.

Ideally this benchmark is representative of a real-world scenarios (that is, care must be taken to ensure that e.g. running with an empty database doesn’t make this benchmark produce numbers better than what they would otherwise be when running on mainnet.) Even better if this benchmark is self-contained, can be run locally without spinning a full network. That would enable use of this benchmark to profile the codebase and narrow down the actionable hotspots if it turns out that the current performance is non-satisfactory.

---

## Writing down a benchmark against real data

We have considered a couple options for making a realistic benchmark for this. An option is to fork mainnet, pick one of larger fungible token transfer contracts and have it do things. We identified a fair number of pain points with this approach, at least one of which is the fact that set-up for such a benchmark is prohibitively expensive. It would also require some ungodly hacks to even start exploring…

So far the consensus seems that it might be better to see where we're spending time and attempt to emulate the relevant portions of the code-base with more self-contained benchmarks.

## Hypotheses

1. Fungible token transfer is a fairly simple contract that does not do very much. There are some potential cost centres we may want to focus on improving

    * Storage operation cost
    * Function call cost
    * WASM execution cost

2. Even if performance of e.g. contract runtime is good enough, improving it further may make it easier to achieve the key result if storage work needs extra time, or vice-versa.
3. We may _already_ have sufficient performance for this, gotta verify this in practice (both in terms of gas amount, and in terms of wall-clock time/instructions)
4. Even if we took one of the actual fungible token transfer contracts and benchmarked it to the ground, the very fact that they are so simple may mean that even minor perturbations to the code logic might shift the equation significantly. We may need a representative set of FT transfer contracts.

### PR DESCRIPTION
**Pull request by nagisa**
I have rewritten the contract since its initial incarnation and it seems to be working quite well so far, so this is probably a good time to land it.

This has also been structured so that its somewhat easier to adjust to different kinds of workloads (socialdb for example) in the future.

cc @akashin @jakmeier 

Fixes #8487

### CODE REVIEW

File: core/o11y/src/lib.rs [hunk 1]
```diff
@@ -205,7 +215,7 @@ where
     S: tracing::Subscriber + for<'span> LookupSpan<'span> + Send + Sync,
     W: for<'writer> fmt::MakeWriter<'writer> + 'static,
 {
-    let layer = fmt::layer().with_ansi(ansi).with_writer(writer).with_filter(filter);
+    let layer = fmt::layer().json().with_ansi(ansi).with_writer(writer).with_filter(filter);
```
[the remark this answers is not in the document]
Comment (nikurt) [reply core/o11y/src/lib.rs:218]: I would really like to have `json` output as an option.
For example, when I was recently dumping all state parts, having json output would have saved me from ~~asking ChatGPT to generate~~ writing a Python script that parses the logs.

**Review by nikurt — COMMENTED** [review]
Looks mostly OK, but I want more comments in the Python files.

File: pytest/tests/stress/perf_ft_transfer.py [hunk 2]
```diff
@@ -0,0 +1,471 @@
...
+```
+"""
+
+import argparse
+import sys
+import os
+import time
+import pathlib
+import base58
+import itertools
```
Comment (marcelo-gonzalez) [inline pytest/tests/stress/perf_ft_transfer.py:27]: could delete the itertools

File: pytest/tests/stress/perf_ft_transfer.py [hunk 3]
```diff
@@ -0,0 +1,471 @@
...
+    except ValueError:
+        for sub in args.contracts.split(","):
+            funding_key = ACCOUNTS[funding_account].key
+            sub_key = key.Key(f"{sub}.{funding_key.account_id}", funding_key.pk, funding_key.sk)
+            contract_accounts.append(len(ACCOUNTS))
+            ACCOUNTS.append(Account(sub_key))
+
+
+    for i in range(int(args.accounts)):
+        keys = ed25519.create_keypair(entropy=rng.randbytes)
```
Comment (marcelo-gonzalez) [inline pytest/tests/stress/perf_ft_transfer.py:402]: This only works for python 3.9 and above. Is overriding the default worth that requirement?

**Review by marcelo-gonzalez — APPROVED** [review]

File: pytest/tests/stress/perf_ft_transfer.py [hunk 4]
```diff
@@ -0,0 +1,471 @@
...
+        else:
+            signer_key = key.Key.from_json_file(args.contract_key)
+
+    else:
+        nodes = [
+            cluster.RpcNode("127.0.0.1", 3030),
+        ]
+        # The `nearup` localnet setup stores the keys in this directory.
+        if args.contract_key is None:
+            key_path = (pathlib.Path.home() / ".near/localnet/node0/shard0_key.json").resolve()
```
Comment (marcelo-gonzalez) [inline pytest/tests/stress/perf_ft_transfer.py:376]: also, was this test script meant to be connected to the other node setup script? Doesn't seem to be possible to use it with that as it is now unless I'm missing something

File: pytest/tests/stress/perf_ft_transfer.py [hunk 5]
```diff
@@ -0,0 +1,471 @@
...
+            "receiver_id": recipient.key.account_id,
+            "amount": str(int(self.how_much)),
+        }
+        tx = transaction.sign_function_call_tx(
+            sender.key,
+            ft.key.account_id,
+            "ft_transfer",
+            json.dumps(args).encode('utf-8'),
+            # About enough gas per call to fit N such transactions into an average block.
+            int(GAS_PER_BLOCK // TRANSACTIONS_PER_BLOCK),
```
Comment (akashin) [inline pytest/tests/stress/perf_ft_transfer.py:162]: I don't think the amount of attached gas has any effect on the number of transactions that will be included in the block as the block fullness is computed based on the burnt gas.

I think we should just use a constant number here instead that is enough to cover the cost of that FT transfer, something like 50TGas should be more than enough.

We can then remove `TRANSACTIONS_PER_BLOCK` constant as it is not used anywhere and is a bit misleading.

File: pytest/tests/stress/perf_ft_transfer.py [hunk 6]
```diff
@@ -0,0 +1,471 @@
...
+ACCOUNTS = []
+
+ID = 0
+
+class Transaction:
+    """
+    A transaction future.
+    """
+    def __init__(self):
+        global ID
```
Comment (akashin) [inline pytest/tests/stress/perf_ft_transfer.py:65]: You can avoid the need for global with a class static variable:
```python
class Test:
  ID = 3

  def __init__(self):
      self.id = Test.ID
      Test.ID += 1

t1 = Test()
print(t1.id)

t2 = Test()
print(t2.id)
```

**Review by akashin — APPROVED** [review]

**Review by nikurt — APPROVED** [review]

File: pytest/tests/stress/perf_ft_transfer.py [hunk 7]
```diff
@@ -0,0 +1,471 @@
+#!/usr/bin/env python3
+
+"""
+This is a benchmark in which a network with a single fungible_token contract is
+deployed, and then a variable number of users (see `N_ACCOUNTS`) send each other
+them fungible tokens.
```
Comment (nikurt) [inline pytest/tests/stress/perf_ft_transfer.py:6]: typo: "each other them"

**Review by nikurt — APPROVED** [review]

**Comment by marcelo-gonzalez** [comment]
> Hm, I've been running this successfully for a while but maybe this comment is an indication that this script is already starting to do too much given that I'm not even testing the many different configurations and there are some modes that make assumptions which don't work out well.

yeah I tested it on my Linux computer at home, and got that error when I tried it on a Mac

File: pytest/tests/stress/perf_ft_transfer.py (hunk 3 above)
Comment (nagisa) [reply pytest/tests/stress/perf_ft_transfer.py:402]: So the intent here was to have at least some semblance of test reproducibility – if something goes wrong, you can specify the same seed as before and get the same implicit accounts for the test.

File: pytest/tests/stress/perf_ft_transfer.py (hunk 4 above)
Comment (nagisa) [reply pytest/tests/stress/perf_ft_transfer.py:376]: This is something @akashin added, I then added the `--contract-key` argument later. I think we can remove this entirely and use `--contract-key` exclusively.

File: pytest/tests/stress/perf_ft_transfer.py (hunk 5 above)
Comment (nagisa) [reply pytest/tests/stress/perf_ft_transfer.py:162]: Fair, I think it might make sense to make this into an CLI argument while I’m changing this.

**Comment by nagisa** [comment]
So, uh, I’m about to look into the macOS behaviour, but it seems like it may be a difference between a `fork(2)`ed process and a `vfork(2)`ed one. If this difference is indeed because of that it would be platform abstraction bug in python, unfortunately.

**Pull request merged**